source/opt/aggressive_dead_code_elim_pass.cpp

```diff
@@ -123,40 +129,193 @@ void AggressiveDCEPass::ProcessLoad(uint32_t varId) {
+    if (IsStructuredIfHeader(*bi, &mergeInst, &branchInst, &mergeBlockId)) {
+      currentMergeBlockId.push(mergeBlockId);
+      currentMergeInst.push(mergeInst);
+      currentBranchInst.push(branchInst);
+    }
+  }
+}
+
+void AggressiveDCEPass::ComputeInst2BlockMap(ir::Function* func) {
+  for (auto& blk : *func) {
+    blk.ForEachInst([&blk,this](ir::Instruction* ip) {
+      inst2block_[ip] = &blk;
+    });
+  }
+}
+
+void AggressiveDCEPass::AddBranch(uint32_t labelId, ir::BasicBlock* bp) {
+  std::unique_ptr<ir::Instruction> newBranch(
+    new ir::Instruction(SpvOpBranch, 0, 0,
+        {{spv_operand_type_t::SPV_OPERAND_TYPE_ID, {labelId}}}));
+  def_use_mgr_->AnalyzeInstDefUse(&*newBranch);
+  bp->AddInstruction(std::move(newBranch));
+}
+
 bool AggressiveDCEPass::AggressiveDCE(ir::Function* func) {
+  // Compute map from instruction to block
+  ComputeInst2BlockMap(func);
+  // Compute map from block to controlling conditional branch
+  std::list<ir::BasicBlock*> structuredOrder;
+  ComputeStructuredOrder(func, &structuredOrder);
+  ComputeBlock2BranchMaps(structuredOrder);
   bool modified = false;
   // Add all control flow and instructions with external side effects 
   // to worklist
   // TODO(greg-lunarg): Handle Frexp, Modf more optimally
   call_in_func_ = false;
   func_is_entry_point_ = false;
   private_stores_.clear();
-  for (auto& blk : *func) {
-    for (auto& inst : blk) {
-      uint32_t op = inst.opcode();
+  std::stack<bool> skip_control_flow;
+  std::stack<uint32_t> currentMergeBlockId;
+  skip_control_flow.push(false);
+  currentMergeBlockId.push(false);
+  for (auto bi = structuredOrder.begin(); bi != structuredOrder.end(); ++bi) {
+    if ((*bi)->id() == currentMergeBlockId.top()) {
+      skip_control_flow.pop();
+      currentMergeBlockId.pop();
+    }
+    for (auto ii = (*bi)->begin(); ii != (*bi)->end(); ++ii) {
+      SpvOp op = ii->opcode();
       switch (op) {
         case SpvOpStore: {
           uint32_t varId;
-          (void) GetPtr(&inst, &varId);
+          (void) GetPtr(&*ii, &varId);
           // Mark stores as live if their variable is not function scope
           // and is not private scope. Remember private stores for possible
           // later inclusion
           if (IsVarOfStorage(varId, SpvStorageClassPrivate))
-            private_stores_.push_back(&inst);
+            private_stores_.push_back(&*ii);
           else if (!IsVarOfStorage(varId, SpvStorageClassFunction))
-            worklist_.push(&inst);
+            AddToWorklist(&*ii);
         } break;
         case SpvOpExtInst: {
           // eg. GLSL frexp, modf
-          if (!IsCombinatorExt(&inst))
-            worklist_.push(&inst);
+          if (!IsCombinatorExt(&*ii))
+            AddToWorklist(&*ii);
+        } break;
+        case SpvOpLoopMerge: {
+          skip_control_flow.push(false);
+          currentMergeBlockId.push(
+              ii->GetSingleWordInOperand(kSelectionMergeMergeBlockIdInIdx));
```

Comment:
This is a LoopMerge, so the fact that kSelectionMergeMergeBlockIdInIdx works is accidental.

Reply:
Yeah, I think that was a cut-and-paste error.